src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs

```diff
@@ -168,6 +168,11 @@ public string TestCode
         /// </summary>
         public List<Func<Solution, ProjectId, Solution>> SolutionTransforms { get; } = new List<Func<Solution, ProjectId, Solution>>();
 
+        /// <summary>
+        /// Gets the syntax trees generated by the source generators during the test run.
+        /// </summary>
+        public ImmutableArray<SyntaxTree> GeneratedTrees { get; private set; } = ImmutableArray<SyntaxTree>.Empty;
```

Comment:
❗ Source generators can run more than once during a test. It's not clear if/when/how this property would be valid during the life of a test. I would expect the following conceptual changes:

1. The state of `AnalyzerTest<T>` should not change during the life of a call to `RunAsync`.
2. The `GeneratedTrees` should not appear as a property on `AnalyzerTest<T>`, since that would make initializing the test instance more confusing (the output property would be showing up as an input property).
3. The manner in which generated trees are provided accounts for the fact that generators may run multiple times during a test.

Reply:
about 1:
it already does:
private readonly ConcurrentBag<Workspace> _workspaces = new ConcurrentBag<Workspace>()

about 2:
It won't, because the setter is private

about 3:
The sample you provided overwrites `GetProjectCompilationAsync()` to get access to the generated sources - but isn't that called also many times, so the problem is exactly the same?

Reply:
Maybe wrapping the call to GenerateSources into a new virtual method could be a good solution?
It would simplify extracting the sources a lot

Reply:
> about 1:
> it already does:
> private readonly ConcurrentBag _workspaces = new ConcurrentBag()

This is true, but also specifically designed to account for an unknown number of workspace creations. This was a bit clumsy in its own addition, and hopefully it could be avoided for future cases.

> about 2: It won't, because the setter is private

`TestState` doesn't have a setter at all. The syntax used to initialize `TestState` is also valid for properties that do not have a public setter.

> about 3: The sample you provided overwrites `GetProjectCompilationAsync()` to get access to the generated sources - but isn't that called also many times, so the problem is exactly the same?

It is, but the path that accesses the generated sources is only executed during specific record cases.

> Maybe wrapping the call to GenerateSources into a new virtual method could be a good solution?
It would simplify extracting the sources a lot

`ApplySourceGeneratorAsync` might be the method to expose. The part I don't like is it's a difficult method to override.